src/2017/subbehaviors.org

```diff
@@ -0,0 +1,150 @@
+* Subbehaviors {#t20173}                                 :docs:
+* Questions about Plays?
+- Topics Covered:
+  - Python
+  - State Machines
+  - Play/Tactic/Skill Structure
+- Last meeting was pretty dense
+  - RoboCup is complicated
+  - If you got it, great
+  - If you didn't, great
+    - We need you to ask questions when you don't understand something
+- And today, we're going into...
+** Additional Information                                             :docs:
+- See the previous slide deck and docs for more information, they should be a pretty comprehensive overview of our play system.
+* Subbehaviors!
+- A behavior is a generic name for any skill, play, or tactic.
+- Last meeting we learned about Plays/Tactics/Skills
+  - How do we put them together?
+- Subbehaviors allow you to reuse our behaviors in many plays.
+  - Example: We have /one/ move play, used by kicker tactics, passing tactics, goalie, etc.
+** Examples of Subbehaviors
+- Offense Play contains Passing Tactics, and Move Skills
+- Passing Tactic Contains Move and Kick Skills
+** /Behavior/ Tree!
+#+BEGIN_SRC text
+Basic122::running
+    Mark::running[robot=4]
+        move(0.767296, 6.51286)
+        endVelocity(0, 0)
+        face(0.160718, 9.15305)
+    PivotKick::capturing
+        Capture::course_approach[robot=2]
+            face(0.160718, 9.15305)
+            avoidBall(-1)
+            move(0.176939, 9.114)
+            endVelocity(0, 0)
+    Mark::running[robot=1]
+        move(-0.829012, 6.03814)
+        endVelocity(0, 0)
+        face(0.160718, 9.15305)
+#+END_SRC
+
+*** Details                                                          :docs:
+- Here we have a tree of all the behaviors running
+- We have a basic122 play running
+- It is running 2 Marks, and a PivotKick in this frame
+- The marks are directly running robot commands, like move, and face.
+- The PivotKick is running a capture skill, which is running robot commands, like face, move, etc.
+** /Behavior/ Tree! (cont.)
+#+BEGIN_SRC text
+Basic122::running
+    Defense::defending
+        SubmissiveDefender::marking
+            Move::completed[robot=5]
+                face(0.160718, 9.15305)
+        SubmissiveDefender::marking
+            Move::completed[robot=3]
+                face(0.160718, 9.15305)
+        SubmissiveGoalie::block
+            Move::running[robot=0]
+                face(0.160718, 9.15305)
+                move(0.290916, 0.14)
+                endVelocity(0, 0)
+#+END_SRC
+
+*** Details                                                          :docs:
+- Pretty much same thing here
+- We've got a Defense tactic running, with SubmissiveDefender and Goalie Skills
+- These skills actually run robot commands.
+* How do I use subbehaviors?
+** Adding Subbehaviors
+#+BEGIN_SRC python
+  self.add_subbehavior(skill_object,
+                       "Name of Subbehavior",
+                       required=False or True
+                       priority=10) # A higher number is higher priority
```

Comment:
Just as a FYI, the `required` and `priority` part is only true if the subbehavior you are adding is not a `complex behavior` or a `single robot complex behavior`.